src/App.tsx

```diff
@@ -1,27 +1,94 @@
-import React from 'react';
+import React, { useState } from 'react';
+import classNames from 'classnames';
+import 'bulma/css/bulma.min.css';
+
+import { Good } from './types/Good';
+import { LoadButtonSettings } from './classes/LoadButtonSettings';
+import { LoadingError } from './components/LoadingError/LoadingError';
+import { GoodsList } from './components/GoodsList/GoodsList';
+import { getAllGoods, getFirstFiveGoods, getRedGoods } from './api/goods';
+
 import './App.scss';
-import { GoodsList } from './GoodsList';
 
-// import { getAll, get5First, getRed } from './api/goods';
-// or
-// import * as goodsAPI from './api/goods';
+const loadButtonsSettings = [
+  new LoadButtonSettings('all', 'all', getAllGoods),
+  new LoadButtonSettings('5 first', 'first-five', getFirstFiveGoods),
+  new LoadButtonSettings('red', 'red', getRedGoods),
+];
+
+export const App: React.FC = () => {
+  const [goods, setGoods] = useState<Good[]>([]);
+  const [isLoading, setIsLoading] = useState(false);
+  const [isLoaded, setIsLoaded] = useState(false);
+  const [isFailedLoad, setIsFailedLoad] = useState(false);
+  const [currentLoadingMode, setCurrentLoadingMode] = useState('');
+
+  const loadGoods = async (
+    callback: () => Promise<Good[]>,
+    namePart: string,
+  ) => {
+    setIsLoaded(false);
+    setIsFailedLoad(false);
+    setCurrentLoadingMode(namePart);
+    setIsLoading(true);
+
+    try {
+      const loadedGoods = await callback();
+
+      if (JSON.stringify(loadedGoods) !== JSON.stringify(goods)) {
+        setGoods(loadedGoods);
+      }
+
+      setIsLoaded(true);
+      setIsLoading(false);
+    } catch (error) {
+      setIsFailedLoad(true);
+      setIsLoading(false);
+    }
+  };
+
+  return (
+    <div className="App">
+      <h1 className="title">
+        Dynamic list of Goods
+      </h1>
 
-export const App: React.FC = () => (
-  <div className="App">
-    <h1>Dynamic list of Goods</h1>
+      <div className="loading-options">
+        {loadButtonsSettings.map(loadButtonSettings => {
+          const {
+            namePart,
+            dataCyPrefix,
+            onClickCallback,
+          } = loadButtonSettings;
 
-    <button type="button" data-cy="all-button">
-      Load all goods
-    </button>
+          const isCurrentMode = namePart === currentLoadingMode;
 
-    <button type="button" data-cy="first-five-button">
-      Load 5 first goods
-    </button>
+          return (
+            <button
+              key={namePart}
+              type="button"
+              className={classNames(
+                'button',
+                'is-info',
+                { 'is-light': !isCurrentMode },
+                { 'is-loading': isCurrentMode && isLoading },
+              )}
+              data-cy={`${dataCyPrefix}-button`}
+              onClick={() => loadGoods(onClickCallback, namePart)}
+            >
+              {`Load ${namePart} goods`}
+            </button>
+          );
+        })}
```

Comment:
Looks complicated. You can create Button component and move this logic into component. And if you do so than you can combine your class LoadButtonSettings with Button component